sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -181,6 +197,12 @@ case class DataSource(
       throw new AnalysisException(
         s"Unable to infer schema for $format. It must be specified manually.")
     }
+
+    checkColumnNameDuplication(
+      (dataSchema ++ partitionSchema).map(_.name),
+      "in the datasource",
+      sparkSession.sessionState.conf.caseSensitiveAnalysis)
```

Comment:
Since the new RC of 2.2 is out, how about submitting the PR to the master branch using your `SchemaUtils.scala` with the corresponding test cases, instead of creating a new function `checkColumnNameDuplication`? 

This can simplify the future back-port merging to 2.2.